File: game.py
import pygame
from player import Player
from obstacle import Obstacle
from trash import Trash
from boost import Boost
from utils import SCREEN_WIDTH, SCREEN_HEIGHT, AUDIO_PATH, LOGO_PATH
from random import randint, choices
import time

from pygame.locals import K_ESCAPE, KEYDOWN, QUIT


class Button:
    def __init__(self, x, y, w, h, text, color=(30, 159, 199), hover=(20, 120, 160)):
        self.rect = pygame.Rect(x, y, w, h)
        self.text = text
        self.color = color
        self.hover = hover

    def draw(self, surface):
        col = self.hover if self.rect.collidepoint(pygame.mouse.get_pos()) else self.color
        pygame.draw.rect(surface, col, self.rect, border_radius=8)
        font = pygame.font.SysFont('Arial', 28)
        txt = font.render(self.text, True, (255, 255, 255))
        surface.blit(txt, txt.get_rect(center=self.rect.center))

    def clicked(self, event):
        return event.type == pygame.MOUSEBUTTONDOWN and event.button == 1 and self.rect.collidepoint(event.pos)


class Toggle:
    def __init__(self, label, x, y, value=True):
        self.label = label
        self.x = x
        self.y = y
        self.value = value
        self.rect = pygame.Rect(x + 280, y - 5, 70, 32)

    def draw(self, surface):
        font = pygame.font.SysFont('Arial', 26)
        surface.blit(font.render(self.label + ':', True, (0, 0, 0)), (self.x, self.y))
        col = (50, 200, 100) if self.value else (180, 60, 60)
        pygame.draw.rect(surface, col, self.rect, border_radius=16)
        val = font.render('ON' if self.value else 'OFF', True, (255, 255, 255))
        surface.blit(val, val.get_rect(center=self.rect.center))

    def handle(self, event):
        if event.type == pygame.MOUSEBUTTONDOWN and event.button == 1 and self.rect.collidepoint(event.pos):
            self.value = not self.value


class Game:
    SETTINGS = 'settings'
    PLAYING = 'playing'
    GAME_OVER = 'game_over'

    def __init__(self):
        pygame.mixer.pre_init(44100, -16, 2, 512)
        pygame.init()
        self.screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
        self.clock = pygame.time.Clock()
        self.logo = pygame.image.load(LOGO_PATH)
        pygame.display.set_caption('Beach Cleanup')
        pygame.display.set_icon(self.logo)
        self.bg = pygame.transform.scale(
            pygame.image.load('./imgs/sand_bg.jpg'), (SCREEN_WIDTH, SCREEN_HEIGHT)
        )

        cx = SCREEN_WIDTH // 2
        self.music_toggle = Toggle('Background Music', cx - 175, SCREEN_HEIGHT // 2 - 10)
        self.sounds_toggle = Toggle('Sound Effects', cx - 175, SCREEN_HEIGHT // 2 + 50)
        self.play_btn = Button(cx - 80, SCREEN_HEIGHT // 2 + 120, 160, 48, 'Play')
        self.play_again_btn = Button(cx - 100, SCREEN_HEIGHT // 2 + 120, 200, 48, 'Play Again')

        self.state = self.SETTINGS
        self.sounds = {}
        self.player = None
        self.obstacles = self.trash = self.boosts = None
        self.startTime = self.totalTime = 0
        self.ADD_OBSTACLE = self.ADD_TRASH = self.ADD_BOOST = None
        self.POINTS_TRASH = 20
        self.POINTS_BOOST = 100
        self.COLLISION_DELTA = 30

    def _init_round(self):
        self.ADD_OBSTACLE = pygame.USEREVENT + 1
        self.ADD_TRASH = pygame.USEREVENT + 2
        self.ADD_BOOST = pygame.USEREVENT + 3
        pygame.time.set_timer(self.ADD_OBSTACLE, randint(1000, 3000))
        pygame.time.set_timer(self.ADD_TRASH, randint(1000, 3000))
        pygame.time.set_timer(self.ADD_BOOST, randint(5000, 10000))

        self.player = Player()
        self.obstacles = pygame.sprite.Group()
        self.trash = pygame.sprite.Group()
        self.boosts = pygame.sprite.Group()
        self.startTime = time.time()
        self.totalTime = 0

        pygame.mixer.music.load(AUDIO_PATH + 'waves.ogg')
        pygame.mixer.music.set_volume(1)
        if self.music_toggle.value:
            pygame.mixer.music.play(loops=-1)

        collect_trash = pygame.mixer.Sound(AUDIO_PATH + 'collect_trash.ogg')
        collect_trash.set_volume(0.8)
        self.sounds['trash'] = collect_trash

        collect_boost = pygame.mixer.Sound(AUDIO_PATH + 'collect_boost.ogg')
        collect_boost.set_volume(0.8)
        self.sounds['boost'] = collect_boost

        game_over = pygame.mixer.Sound(AUDIO_PATH + 'game_over.ogg')
        game_over.set_volume(0.2)
        self.sounds['game_over'] = game_over

        self.state = self.PLAYING

    # ── Settings screen ──────────────────────────────────────────────

    def _draw_settings(self):
        self.screen.fill((240, 220, 160))
        logo_scaled = pygame.transform.scale(self.logo, (220, 110))
        self.screen.blit(logo_scaled, logo_scaled.get_rect(center=(SCREEN_WIDTH // 2, SCREEN_HEIGHT // 4)))
        font = pygame.font.SysFont('Arial', 38, bold=True)
        title = font.render('Settings', True, (0, 0, 0))
        self.screen.blit(title, title.get_rect(center=(SCREEN_WIDTH // 2, SCREEN_HEIGHT // 2 - 65)))
        self.music_toggle.draw(self.screen)
        self.sounds_toggle.draw(self.screen)
        self.play_btn.draw(self.screen)

    def _handle_settings(self, event):
        self.music_toggle.handle(event)
        self.sounds_toggle.handle(event)
        if self.play_btn.clicked(event):
            self._init_round()

    # ── Playing screen ───────────────────────────────────────────────

    def _collision(self, a, b):
        x_ok = (a.rect.right > b.rect.left) and ((a.rect.right - b.rect.left) >= self.COLLISION_DELTA)
        top = (a.rect.bottom > b.rect.top) and (self.COLLISION_DELTA <= (a.rect.bottom - b.rect.top) <= (a.rect.height - self.COLLISION_DELTA))
        bot = (b.rect.bottom > a.rect.top) and (self.COLLISION_DELTA <= (b.rect.bottom - a.rect.top) <= (a.rect.height - self.COLLISION_DELTA))
        return x_ok and (top or bot)

    def _check_collisions(self):
        for obs in self.obstacles:
            if self._collision(self.player, obs):
                if self.sounds_toggle.value:
                    self.sounds['game_over'].play()
                pygame.mixer.music.stop()
                self.state = self.GAME_OVER
                return

        collected_trash = pygame.sprite.spritecollide(self.player, self.trash, True)
        if collected_trash:
            self.player.incPts(self.POINTS_TRASH * len(collected_trash))
            if self.sounds_toggle.value:
                self.sounds['trash'].play()

        collected_boosts = pygame.sprite.spritecollide(self.player, self.boosts, True)
        if collected_boosts:
            self.player.incPts(self.POINTS_BOOST * len(collected_boosts))
            if self.sounds_toggle.value:
                self.sounds['boost'].play()

    def _draw_game(self):
        self.screen.blit(self.bg, (0, 0))
        for sprite in list(self.obstacles) + list(self.trash) + list(self.boosts) + [self.player]:
            self.screen.blit(sprite.surface, sprite.rect)
        font = pygame.font.SysFont('Arial', 50)
        score = font.render('Score: ' + str(self.player.getPts()), True, (0, 0, 0))
        self.screen.blit(score, score.get_rect(topright=(SCREEN_WIDTH - 10, 15)))
        timer = font.render('Time: ' + str(self.totalTime) + 's', True, (0, 0, 0))
        self.screen.blit(timer, timer.get_rect(bottomright=(SCREEN_WIDTH - 10, SCREEN_HEIGHT - 10)))

    def _handle_game(self, event):
        if event.type == KEYDOWN and event.key == K_ESCAPE:
            pygame.mixer.music.stop()
            self.state = self.GAME_OVER
        elif event.type == self.ADD_OBSTACLE:
            obj = Obstacle()
            self.obstacles.add(obj)
        elif event.type == self.ADD_TRASH:
            obj = Trash()
            self.trash.add(obj)
        elif event.type == self.ADD_BOOST:
            if choices([0, 1], weights=[70, 30])[0]:
                self.boosts.add(Boost())

    def _update_game(self):
        keys = pygame.key.get_pressed()
        self.player.update(keys)
        self.obstacles.update()
        self.trash.update()
        self.boosts.update()
        self.totalTime = int(time.time() - self.startTime)
        self._check_collisions()

    # ── Game over screen ─────────────────────────────────────────────

    def _draw_game_over(self):
        self.screen.blit(self.bg, (0, 0))
        overlay = pygame.Surface((SCREEN_WIDTH, SCREEN_HEIGHT), pygame.SRCALPHA)
        overlay.fill((0, 0, 0, 160))
        self.screen.blit(overlay, (0, 0))

        cx = SCREEN_WIDTH // 2
        big = pygame.font.SysFont('Arial', 60, bold=True)
        med = pygame.font.SysFont('Arial', 36)

        title = big.render('GAME OVER', True, (255, 255, 255))
        self.screen.blit(title, title.get_rect(center=(cx, SCREEN_HEIGHT // 2 - 90)))
        score = med.render('Score: ' + str(self.player.getPts()), True, (255, 255, 255))
        self.screen.blit(score, score.get_rect(center=(cx, SCREEN_HEIGHT // 2 - 10)))
        t = med.render('Time: ' + str(self.totalTime) + 's', True, (255, 255, 255))
        self.screen.blit(t, t.get_rect(center=(cx, SCREEN_HEIGHT // 2 + 45)))
        self.play_again_btn.draw(self.screen)

    def _handle_game_over(self, event):
        if self.play_again_btn.clicked(event):
            self.state = self.SETTINGS

    # ── Main tick (called every frame from run.py) ───────────────────

    def tick(self):
        for event in pygame.event.get():
            if event.type == QUIT:
                return False
            if self.state == self.SETTINGS:
                self._handle_settings(event)
            elif self.state == self.PLAYING:
                self._handle_game(event)
            elif self.state == self.GAME_OVER:
                self._handle_game_over(event)

        if self.state == self.PLAYING:
            self._update_game()

        if self.state == self.SETTINGS:
            self._draw_settings()
        elif self.state == self.PLAYING:
            self._draw_game()
        elif self.state == self.GAME_OVER:
            self._draw_game_over()

        pygame.display.flip()
        self.clock.tick(30)
        return True

File: game.js
const W = 800;
const H = 600;
const COLLISION_DELTA = 30;
const FRAME_MS = 1000 / 30; // speeds are in px/frame at 30fps; scale by delta

const OBSTACLE_KEYS = ['crab', 'seagull', 'seaweed'];
const TRASH_KEYS    = ['can', 'cigarette', 'paper_wad', 'water_bottle'];
const BOOST_KEYS    = ['seashell1', 'seashell2', 'seashell3'];

function rand(min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function hit(x, y, r) {
  return x >= r.x && x <= r.x + r.w && y >= r.y && y <= r.y + r.h;
}

class Game {
  constructor(canvas) {
    this.canvas = canvas;
    this.ctx    = canvas.getContext('2d');

    this.images = {};
    this.sounds = {};
    this.keys   = {};
    this.mouse  = { x: 0, y: 0 };

    this.state      = 'loading';
    this.playMusic  = true;
    this.playSounds = true;

    this.player    = null;
    this.obstacles = [];
    this.trash     = [];
    this.boosts    = [];
    this.score     = 0;
    this.startTime = 0;
    this.elapsed   = 0;

    this.nextObstacleAt = 0;
    this.nextTrashAt    = 0;
    this.nextBoostAt    = 0;

    this.lastTime = 0;

    const cx = W / 2;
    this.ui = {
      musicToggle: { x: cx - 175, y: H / 2 - 10,  w: 370, h: 35,  label: 'Background Music', value: true  },
      soundToggle: { x: cx - 175, y: H / 2 + 50,  w: 370, h: 35,  label: 'Sound Effects',    value: true  },
      play:        { x: cx - 80,  y: H / 2 + 120, w: 160, h: 48,  label: 'Play'                           },
      playAgain:   { x: cx - 100, y: H / 2 + 120, w: 200, h: 48,  label: 'Play Again'                     },
    };

    this._setupListeners();
    this._loadAssets().then(() => {
      this.state = 'settings';
      requestAnimationFrame(t => this._loop(t));
    });
  }

  // ── Asset loading ─────────────────────────────────────────────────

  async _loadAssets() {
    const imgMap = {
      player:       'imgs/players/stick_figure.png',
      bg:           'imgs/sand_bg.jpg',
      logo:         'imgs/logo.png',
      crab:         'imgs/obstacles/crab.png',
      seagull:      'imgs/obstacles/seagull.png',
      seaweed:      'imgs/obstacles/seaweed.png',
      can:          'imgs/trash/can.png',
      cigarette:    'imgs/trash/cigarette.png',
      paper_wad:    'imgs/trash/paper_wad.png',
      water_bottle: 'imgs/trash/water_bottle.png',
      seashell1:    'imgs/boosts/seashell1.png',
      seashell2:    'imgs/boosts/seashell2.png',
      seashell3:    'imgs/boosts/seashell3.png',
    };

    await Promise.all(Object.entries(imgMap).map(([key, src]) =>
      new Promise(resolve => {
        const img = new Image();
        img.onload  = () => { this.images[key] = img; resolve(); };
        img.onerror = () => resolve(); // skip missing images gracefully
        img.src = src;
      })
    ));

    const audioMap = {
      waves:         { src: 'audio/waves.ogg',         loop: true,  volume: 1.0 },
      collect_trash: { src: 'audio/collect_trash.ogg', loop: false, volume: 0.8 },
      collect_boost: { src: 'audio/collect_boost.ogg', loop: false, volume: 0.8 },
      game_over:     { src: 'audio/game_over.ogg',     loop: false, volume: 0.2 },
    };

    for (const [key, { src, loop, volume }] of Object.entries(audioMap)) {
      const a = new Audio(src);
      a.loop   = loop;
      a.volume = volume;
      this.sounds[key] = a;
    }
  }

  // ── Input ─────────────────────────────────────────────────────────

  _setupListeners() {
    document.addEventListener('keydown', e => {
      this.keys[e.key] = true;
      if (['ArrowUp','ArrowDown','ArrowLeft','ArrowRight',' '].includes(e.key)) {
        e.preventDefault();
      }
    });
    document.addEventListener('keyup', e => { this.keys[e.key] = false; });

    this.canvas.addEventListener('mousemove', e => {
      const r = this.canvas.getBoundingClientRect();
      this.mouse.x = (e.clientX - r.left) * (W / r.width);
      this.mouse.y = (e.clientY - r.top)  * (H / r.height);
    });

    this.canvas.addEventListener('click', e => {
      const r = this.canvas.getBoundingClientRect();
      const x = (e.clientX - r.left) * (W / r.width);
      const y = (e.clientY - r.top)  * (H / r.height);
      this._handleClick(x, y);
    });
  }

  _handleClick(x, y) {
    if (this.state === 'settings') {
      const { musicToggle, soundToggle, play } = this.ui;
      if (hit(x, y, musicToggle)) { musicToggle.value = !musicToggle.value; this.playMusic  = musicToggle.value; }
      if (hit(x, y, soundToggle)) { soundToggle.value = !soundToggle.value; this.playSounds = soundToggle.value; }
      if (hit(x, y, play))        this._startGame();
    } else if (this.state === 'gameover') {
      if (hit(x, y, this.ui.playAgain)) {
        this._stopMusic();
        this.state = 'settings';
      }
    }
  }

  // ── Game lifecycle ────────────────────────────────────────────────

  _startGame() {
    this.score     = 0;
    this.startTime = performance.now();
    this.elapsed   = 0;
    this.obstacles = [];
    this.trash     = [];
    this.boosts    = [];

    const pw = W / 8, ph = H / 4;
    this.player = { img: 'player', x: 0, y: H / 2 - ph / 2, w: pw, h: ph };

    const now = performance.now();
    this.nextObstacleAt = now + rand(1000, 3000);
    this.nextTrashAt    = now + rand(1000, 3000);
    this.nextBoostAt    = now + rand(5000, 10000);

    if (this.playMusic) this._playMusic();
    this.state = 'playing';
  }

  _spawnObstacle() {
    const key = OBSTACLE_KEYS[rand(0, OBSTACLE_KEYS.length - 1)];
    return { img: key, x: W + rand(20, 100), y: rand(0, H - H/8), w: W/9, h: H/8, speed: rand(4, 10) };
  }

  _spawnTrash() {
    const key = TRASH_KEYS[rand(0, TRASH_KEYS.length - 1)];
    return { img: key, x: W + rand(20, 100), y: rand(0, H - H/10), w: W/10, h: H/10, speed: rand(4, 8) };
  }

  _spawnBoost() {
    const key = BOOST_KEYS[rand(0, BOOST_KEYS.length - 1)];
    return { img: key, x: W + rand(20, 100), y: rand(0, H - H/10), w: W/10, h: H/10, speed: 3 };
  }

  // ── Collision ─────────────────────────────────────────────────────

  _collides(a, b) {
    const xOverlap = (a.x + a.w) - b.x;
    if (xOverlap < COLLISION_DELTA) return false;

    const topOverlap = (a.y + a.h) - b.y;
    const botOverlap = (b.y + b.h) - a.y;
    const topOk = topOverlap >= COLLISION_DELTA && topOverlap <= a.h - COLLISION_DELTA;
    const botOk = botOverlap >= COLLISION_DELTA && botOverlap <= a.h - COLLISION_DELTA;
    return topOk || botOk;
  }

  // ── Update ────────────────────────────────────────────────────────

  _update(now, delta) {
    const scale = delta / FRAME_MS;
    const p = this.player;

    // Spawn
    if (now >= this.nextObstacleAt) { this.obstacles.push(this._spawnObstacle()); this.nextObstacleAt = now + rand(1000, 3000); }
    if (now >= this.nextTrashAt)    { this.trash.push(this._spawnTrash());         this.nextTrashAt    = now + rand(1000, 3000); }
    if (now >= this.nextBoostAt) {
      if (Math.random() < 0.3) this.boosts.push(this._spawnBoost());
      this.nextBoostAt = now + rand(5000, 10000);
    }

    // Player movement
    if (this.keys['ArrowUp'])    p.y -= 5 * scale;
    if (this.keys['ArrowDown'])  p.y += 5 * scale;
    if (this.keys['ArrowLeft'])  p.x -= 5 * scale;
    if (this.keys['ArrowRight']) p.x += 5 * scale;
    p.x = Math.max(0, Math.min(W - p.w, p.x));
    p.y = Math.max(0, Math.min(H - p.h, p.y));

    // Move sprites and cull off-screen
    for (const s of [...this.obstacles, ...this.trash, ...this.boosts]) s.x -= s.speed * scale;
    this.obstacles = this.obstacles.filter(s => s.x + s.w > 0);
    this.trash     = this.trash.filter(s => s.x + s.w > 0);
    this.boosts    = this.boosts.filter(s => s.x + s.w > 0);

    // Obstacle collision → game over
    for (const obs of this.obstacles) {
      if (this._collides(p, obs)) {
        this.elapsed = Math.floor((now - this.startTime) / 1000);
        if (this.playSounds) this._playSound('game_over');
        this._stopMusic();
        this.state = 'gameover';
        return;
      }
    }

    // Trash collection
    this.trash = this.trash.filter(t => {
      if (this._collides(p, t)) {
        this.score += 20;
        if (this.playSounds) this._playSound('collect_trash');
        return false;
      }
      return true;
    });

    // Boost collection
    this.boosts = this.boosts.filter(b => {
      if (this._collides(p, b)) {
        this.score += 100;
        if (this.playSounds) this._playSound('collect_boost');
        return false;
      }
      return true;
    });

    this.elapsed = Math.floor((now - this.startTime) / 1000);
  }

  // ── Draw ──────────────────────────────────────────────────────────

  _drawBtn(r, label, hoverColor = '#1e9fc7', baseColor = '#1a8db0') {
    const mx = this.mouse.x, my = this.mouse.y;
    const isHover = hit(mx, my, r);
    this.ctx.fillStyle = isHover ? hoverColor : baseColor;
    this.ctx.beginPath();
    this.ctx.roundRect(r.x, r.y, r.w, r.h, 8);
    this.ctx.fill();
    this.ctx.fillStyle = '#fff';
    this.ctx.font = '22px Arial';
    this.ctx.textAlign = 'center';
    this.ctx.textBaseline = 'middle';
    this.ctx.fillText(label, r.x + r.w / 2, r.y + r.h / 2);
  }

  _drawToggle(t) {
    const ctx = this.ctx;
    ctx.fillStyle = '#000';
    ctx.font = '22px Arial';
    ctx.textAlign = 'left';
    ctx.textBaseline = 'middle';
    ctx.fillText(t.label + ':', t.x, t.y + 17);

    const bx = t.x + 285, by = t.y, bw = 70, bh = 32;
    ctx.fillStyle = t.value ? '#32c864' : '#b43c3c';
    ctx.beginPath();
    ctx.roundRect(bx, by, bw, bh, 16);
    ctx.fill();
    ctx.fillStyle = '#fff';
    ctx.font = '17px Arial';
    ctx.textAlign = 'center';
    ctx.textBaseline = 'middle';
    ctx.fillText(t.value ? 'ON' : 'OFF', bx + bw / 2, by + bh / 2);
  }

  _drawSettings() {
    const ctx = this.ctx, cx = W / 2;

    ctx.fillStyle = '#f0dca0';
    ctx.fillRect(0, 0, W, H);

    if (this.images.logo) {
      const lw = 220, lh = 110;
      ctx.drawImage(this.images.logo, cx - lw / 2, H / 4 - lh / 2, lw, lh);
    }

    ctx.fillStyle = '#000';
    ctx.font = 'bold 36px Arial';
    ctx.textAlign = 'center';
    ctx.textBaseline = 'middle';
    ctx.fillText('Settings', cx, H / 2 - 65);

    this._drawToggle(this.ui.musicToggle);
    this._drawToggle(this.ui.soundToggle);
    this._drawBtn(this.ui.play, 'Play');
  }

  _drawGame() {
    const ctx = this.ctx;

    if (this.images.bg) {
      ctx.drawImage(this.images.bg, 0, 0, W, H);
    } else {
      ctx.fillStyle = '#f0dca0';
      ctx.fillRect(0, 0, W, H);
    }

    // Draw sprites: obstacles, trash, boosts, then player on top
    for (const s of [...this.obstacles, ...this.trash, ...this.boosts, this.player]) {
      const img = this.images[s.img];
      if (img) ctx.drawImage(img, s.x, s.y, s.w, s.h);
    }

    // Score (top right)
    ctx.fillStyle = 'rgba(0,0,0,0.55)';
    ctx.font = 'bold 36px Arial';
    ctx.textAlign = 'right';
    ctx.textBaseline = 'top';
    ctx.fillText('Score: ' + this.score, W - 10, 12);

    // Timer (bottom right)
    ctx.textBaseline = 'bottom';
    ctx.fillText('Time: ' + this.elapsed + 's', W - 10, H - 10);
  }

  _drawGameOver() {
    this._drawGame();

    const ctx = this.ctx, cx = W / 2;
    ctx.fillStyle = 'rgba(0,0,0,0.65)';
    ctx.fillRect(0, 0, W, H);

    ctx.fillStyle = '#fff';
    ctx.textAlign = 'center';

    ctx.font = 'bold 58px Arial';
    ctx.textBaseline = 'middle';
    ctx.fillText('GAME OVER', cx, H / 2 - 90);

    ctx.font = '34px Arial';
    ctx.fillText('Score: ' + this.score,         cx, H / 2 - 10);
    ctx.fillText('Time: '  + this.elapsed + 's', cx, H / 2 + 45);

    this._drawBtn(this.ui.playAgain, 'Play Again');
  }

  // ── Audio ─────────────────────────────────────────────────────────

  _playMusic() {
    const s = this.sounds.waves;
    if (s) { s.currentTime = 0; s.play().catch(() => {}); }
  }

  _stopMusic() {
    const s = this.sounds.waves;
    if (s) s.pause();
  }

  _playSound(name) {
    const s = this.sounds[name];
    if (s) { s.currentTime = 0; s.play().catch(() => {}); }
  }

  // ── Loop ──────────────────────────────────────────────────────────

  _loop(timestamp) {
    const delta = Math.min(timestamp - this.lastTime, 100); // cap at 100ms to avoid spiral of death
    this.lastTime = timestamp;

    if (this.state === 'playing') this._update(timestamp, delta);

    this.ctx.clearRect(0, 0, W, H);
    if (this.state === 'settings')  this._drawSettings();
    if (this.state === 'playing')   this._drawGame();
    if (this.state === 'gameover')  this._drawGameOver();

    requestAnimationFrame(t => this._loop(t));
  }
}

window.addEventListener('load', () => new Game(document.getElementById('gameCanvas')));
